### PR TITLE
[WIP] Tornado50

### PR DESCRIPTION
**Pull request by DmitryKuzmenko**
### What does this PR do?
Work is in progress, please ignore this for now.

### What issues does this PR fix or reference?
Fix https://github.com/saltstack/salt-jenkins/issues/995

### Previous Behavior
Salt is non-functional on Python 3 with Tornado 5.0 

### New Behavior
Work in progress but at the end it should work better than before =)

### Tests written?
No

### Commits signed with GPG?
Yes

### CODE REVIEW

File: salt/config/__init__.py [hunk 1]
```diff
@@ -1241,6 +1243,7 @@ def _gather_buffer_space():
     'id_function': {},
     'cachedir': os.path.join(salt.syspaths.CACHE_DIR, 'minion'),
     'append_minionid_config_dirs': [],
+    'worker_threads': 3,
```
Comment (cachedout) [inline salt/config/__init__.py]: Are we certain about this? This default has been `5` for a long time. Could you explain this change more, please?

File: salt/grains/zfs.py [hunk 2]
```diff
@@ -70,11 +70,12 @@ def _zfs_pool_data():
     return grains
 
 
-def zfs():
+def zfs(grains=None):
```
Comment (cachedout) [inline salt/grains/zfs.py]: I'm not following how this change is related to Tornado. Could you clarify, please?

File: salt/key.py [hunk 3]
```diff
@@ -373,7 +373,7 @@ class Key(object):
     REJ = 'minions_rejected'
     DEN = 'minions_denied'
 
-    def __init__(self, opts, io_loop=None):
+    def __init__(self, opts):
```
Comment (cachedout) [inline salt/key.py]: The reason we had this was to correct https://github.com/saltstack/salt/issues/38209

Are we certain that we won't regress here?

**Comment by cachedout** [comment]
We should have @skizunov come by and look at this if he is available to as well.

File: salt/config/__init__.py (hunk 1 above)
Comment (DmitryKuzmenko) [reply salt/config/__init__.py]: It was so for master I'm not sure what's the best one for minion. I would change it to 5.

File: salt/grains/zfs.py (hunk 2 above)
Comment (DmitryKuzmenko) [reply salt/grains/zfs.py]: There was a bug in the beginning of my work that I've fixed this way. I'll check how's it going now and will remove this if it's not needed anymore.

File: salt/key.py (hunk 3 above)
Comment (DmitryKuzmenko) [reply salt/key.py]: I'll check this case when I'll done here.
Anyway Tornado doesn't allow to play with IOLoops now and we shall use the current one everywhere.

**Comment by rallytime** [comment]
This will need to be rebased as well when it's ready as there are some merge conflicts here now.

File: salt/key.py (hunk 3 above)
Comment (isbm) [reply salt/key.py]: > Anyway Tornado doesn't allow to play with IOLoops now and we shall use the current one everywhere.

But we should have the ability to use older ones.

File: salt/minion.py [hunk 4]
```diff
@@ -1480,37 +1495,43 @@ def _handle_decoded_payload(self, data):
...
-                name=data['jid']
-            )
-
-        if multiprocessing_enabled:
-            with default_signals(signal.SIGINT, signal.SIGTERM):
-                # Reset current signals before starting the process in
-                # order not to inherit the current signal handlers
-                process.start()
+        if self.opts.get('ipc_mode', '') == 'tcp':
+            pull_uri = int(self.opts.get('tcp_minion_workers', 4516))
```
Comment (isbm) [inline salt/minion.py]: I think ports needs to be moved somewhere to `salt.utils.default.ports` or something similar. Just like we have exitcodes.

**Review by isbm — COMMENTED** [review]
Keep in mind that we still need Tornado 4.x support backward compat.

File: salt/minion.py (hunk 4 above)
Comment (DmitryKuzmenko) [reply salt/minion.py]: Agree.

File: salt/key.py (hunk 3 above)
Comment (DmitryKuzmenko) [reply salt/key.py]: I understand this, but If we use the only on IOLoop per thread/process, we should be OK with Tornado 4.
Anyway I'll take an additional attention on this.

Comment (isbm) [reply salt/key.py]: @DmitryKuzmenko yeah, I think so too. But just keep in mind that. Sometimes you can have pretty old systems in the enterprise, but who want to have latest possible codebase (with outdated dependencies, of course)... :wink:

**Comment by brejoc** [comment]
@DmitryKuzmenko Is there something we might try to help you with here? I'm asking because this is currently blocking the update to Python 3.7 in openSUSE Tumbleweed.

**Comment by sysrich** [comment]
The lack of python 3.7 support in saltstack is leading to the consideration of salt being removed from openSUSE Tumbleweed and openSUSE Kubic - this is something which I don't want to see from both a personal and obviously political perspective, so if there is anything that can be done to expedite a solution I would greatly appreciate it.

**Comment by gtmanfred** [comment]
You can use python 3.7 with 2018.3.3 and 2017.7.8.

You just cannot use tornado 5.0.

**Comment by dwoz** [comment]
I've pointed this to the tornado50 branch. I'm wondering if we should close this issue (for now) and pull stuff into the tornado50 branch from here when needed?

**Comment by Ch3LL** [comment]
@DmitryKuzmenko bump on @dwoz question above

**Comment by DmitryKuzmenko** [comment]
@Ch3LL sorry, thanks for mention.
@dwoz yes I agree.

**Pull request closed without merging**